src/khepri_projection.erl

```diff
@@ -0,0 +1,296 @@
+%%
+%% In some cases, a node in the store might correspond to many objects in a
+%% projection table. Extended projection functions are allowed to call ETS
+%% functions directly in order to build the projection table.
+%%
+%% `OldPayload' and `NewPayload' are undefined if there is no payload for the
+%% node.
+%%
+%% This function is compiled like a transaction function except that calls
+%% to the `ets' module are allowed.
+
+-type projection_fun() :: simple_projection_fun() | extended_projection_fun().
+%% A function that formats an entry in the tree into a record to be stored
+%% in a projection.
+
+-type options() :: #{type => ets:table_type(),
+                     keypos => pos_integer(),
+                     read_concurrency => boolean(),
+                     write_concurrency => boolean() | auto,
+                     compressed => boolean()}.
+%% Options which control the created ETS table.
+%%
+%% These options are a subset of the options available to {@link ets:new/2}.
+%% Refer to the {@link ets:new/2} documentation for a reference on each type
+%% and available values.
+
+-spec new(Name, ProjectFun) -> Projection when
+      Name :: atom(),
+      ProjectFun :: projection_fun(),
+      Projection :: projection().
+%% Creates a new projection data structure with default options.
+%% This is the same as calling `new(Name, ProjectFun, [])'.
+
+new(Name, ProjectFun) ->
+    new(Name, ProjectFun, #{}).
+
+-spec new(Name, ProjectFun, Options) -> Projection when
+      Name :: atom(),
+      ProjectFun :: projection_fun(),
+      Options :: options(),
+      Projection :: projection().
+%% @doc Creates a new projection data structure.
+%%
+%% @param Name the name of the projection. This corresponds to the name of
+%%        the ETS table which is created when the projection is registered.
+%% @param ProjectFun the function which turns paths and data into records to
+%%        be stored in the projection table.
+%% @param Options options which control properties of the projection table.
+%%
+%% @returns a {@link projection()} resource. An error is thrown in the shape
+%%          of `{unexpected_option, Key, Value}' when an unknown or invalid
+%%          {@link option()} is passed.
+
+
+new(Name, Fun, Options)
+  when is_map(Options) andalso
+       (is_function(Fun, 2) orelse is_function(Fun, 4)) ->
+    EtsOptions = maps:fold(fun to_ets_options/3, [named_table], Options),
+    ProjectFun =
+    if
+        is_function(Fun, 2) ->
+            FunOptions = #{ensure_instruction_is_permitted =>
+                           fun khepri_tx_adv:ensure_instruction_is_permitted/1,
+                           should_process_function =>
+                           fun khepri_tx_adv:should_process_function/4,
+                           is_standalone_fun_still_needed =>
+                           fun(_Params) -> true end},
+            StandaloneFun = khepri_fun:to_standalone_fun(Fun, FunOptions),
+            {simple, StandaloneFun};
```

Comment:
You could still query the arity of the extracted function later. I wonder if it's necessary to wrap the function in a `{simpl | extended, StandaloneFun}` tuple. Perhaps it's more future proof.